File: src/main/java/io/polygloat/model/Translation.java
package io.polygloat.model;

import javax.persistence.*;

@Entity
@Table(uniqueConstraints = {
        @UniqueConstraint(columnNames = {"key_id", "language_id"}, name = "translation_key_language"),
})
public class Translation extends AuditModel {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(columnDefinition = "text")
    private String text;

    @ManyToOne
    private Key key;

    @ManyToOne
    private Language language;

    public Translation(Long id, String text, Key key, Language language) {
        this.id = id;
        this.text = text;
        this.key = key;
        this.language = language;
    }

    public Translation() {
    }

    public static TranslationBuilder builder() {
        return new TranslationBuilder();
    }

    public Language getLanguage() {
        return language;
    }

    public Key getKey() {
        return key;
    }

    public String getText() {
        return text;
    }

    public void setKey(Key key) {
        this.key = key;
    }

    public void setLanguage(Language language) {
        this.language = language;
    }

    public void setText(String text) {
        this.text = text;
    }

    public Long getId() {
        return this.id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String toString() {
        return "Translation(id=" + this.getId() + ", text=" + this.getText() + ", key=" + this.getKey() + ", language=" + this.getLanguage() + ")";
    }

    public boolean equals(final Object o) {
        if (o == this) return true;
        if (!(o instanceof Translation)) return false;
        final Translation other = (Translation) o;
        if (!other.canEqual((Object) this)) return false;
        if (!super.equals(o)) return false;
        final Object this$id = this.getId();
        final Object other$id = other.getId();
        if (this$id == null ? other$id != null : !this$id.equals(other$id)) return false;
        final Object this$text = this.getText();
        final Object other$text = other.getText();
        if (this$text == null ? other$text != null : !this$text.equals(other$text)) return false;
        return true;
    }

    protected boolean canEqual(final Object other) {
        return other instanceof Translation;
    }

    public int hashCode() {
        final int PRIME = 59;
        int result = super.hashCode();
        final Object $id = this.getId();
        result = result * PRIME + ($id == null ? 43 : $id.hashCode());
        final Object $text = this.getText();
        result = result * PRIME + ($text == null ? 43 : $text.hashCode());
        return result;
    }

    public static class TranslationBuilder {
        private Long id;
        private String text;
        private Key key;
        private Language language;

        TranslationBuilder() {
        }

        public Translation.TranslationBuilder id(Long id) {
            this.id = id;
            return this;
        }

        public Translation.TranslationBuilder text(String text) {
            this.text = text;
            return this;
        }

        public Translation.TranslationBuilder key(Key key) {
            this.key = key;
            return this;
        }

        public Translation.TranslationBuilder language(Language language) {
            this.language = language;
            return this;
        }

        public Translation build() {
            return new Translation(id, text, key, language);
        }

        public String toString() {
            return "Translation.TranslationBuilder(id=" + this.id + ", text=" + this.text + ", key=" + this.key + ", language=" + this.language + ")";
        }
    }
}

File: src/main/kotlin/io/tolgee/model/Translation.kt
package io.tolgee.model

import io.tolgee.service.dataImport.ImportService
import org.springframework.beans.factory.ObjectFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Configurable
import org.springframework.transaction.annotation.Transactional
import javax.persistence.*

@Entity
@EntityListeners(Translation.Companion.TranslationListeners::class)
@Table(
        uniqueConstraints = [UniqueConstraint(
                columnNames = ["key_id", "language_id"],
                name = "translation_key_language"
        )]
)
data class Translation(
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        var id: Long? = null,

        @Column(columnDefinition = "text")
        var text: String? = null
) : AuditModel() {


    @ManyToOne
    var key: Key? = null

    @ManyToOne
    var language: Language? = null

    constructor(id: Long?, text: String?, key: Key?, language: Language?) : this(id, text) {
        this.key = key
        this.language = language
    }


    class TranslationBuilder internal constructor() {
        private var id: Long? = null
        private var text: String? = null
        private var key: Key? = null
        private var language: Language? = null
        fun id(id: Long?): TranslationBuilder {
            this.id = id
            return this
        }

        fun text(text: String?): TranslationBuilder {
            this.text = text
            return this
        }

        fun key(key: Key?): TranslationBuilder {
            this.key = key
            return this
        }

        fun language(language: Language?): TranslationBuilder {
            this.language = language
            return this
        }

        fun build(): Translation {
            return Translation(id, text, key, language)
        }

        override fun toString(): String {
            return "Translation.TranslationBuilder(id=$id, text=$text, key=$key, language=$language)"
        }
    }

    companion object {
        @JvmStatic
        fun builder(): TranslationBuilder {
            return TranslationBuilder()
        }

        @Configurable
        open class TranslationListeners {

            @Autowired
            open lateinit var provider: ObjectFactory<ImportService>

            @PreRemove
            @Transactional
            open fun preRemove(translation: Translation) {
                provider.`object`.onTranslationConflictRemoved(translation)
            }
        }
    }
}
